### PR TITLE
[Materials] Add parsing support for blur and vibrancy effects

### PR DESCRIPTION
**Pull request by pxlcoder**
#### 3ac783ab6d8f72105ed2463ea9368e09887be975
<pre>
[Materials] Add parsing support for blur and vibrancy effects
<a href="https://bugs.webkit.org/show_bug.cgi?id=284720">https://bugs.webkit.org/show_bug.cgi?id=284720</a>
<a href="https://rdar.apple.com/141519576">rdar://141519576</a>

Reviewed by Abrar Rahman Protyasha, Wenson Hsieh, and Tim Nguyen.

Work towards an implementation of system materials in WebKit. This is not
intended to be web-exposed and is only for internal clients.

The API surface is intended to mirror platform visual effect views, which
support both blur and vibrancy effects. At most one effect can be specified at
a time, via `-apple-visual-effect`.

See <a href="https://developer.apple.com/design/human-interface-guidelines/materials">https://developer.apple.com/design/human-interface-guidelines/materials</a>
for information about system materials.

* LayoutTests/TestExpectations:
* LayoutTests/apple-visual-effects/apple-visual-effect-parsing-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-parsing.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/platform/cocoa/AppleVisualEffect.cpp: Added.
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/cocoa/AppleVisualEffect.h: Added.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp: Unified sources build fix.
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp: Unified sources build fix.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::appleVisualEffect const):
(WebCore::RenderStyle::inheritedCustomProperties const):
(WebCore::RenderStyle::initialAnchorNames):
(WebCore::RenderStyle::initialAppearance):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setAppleVisualEffect):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:

Canonical link: <a href="https://commits.webkit.org/287975@main">https://commits.webkit.org/287975@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9e19a169784d8d998bfe056765341f535e0cfac6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1005 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86009 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32474 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83590 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8824 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63600 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21342 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84549 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/717 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74157 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43891 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28322 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30924 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74459 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28912 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87444 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80538 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8710 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6191 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71919 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8891 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69978 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71150 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15215 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14132 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102947 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12637 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8672 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14201 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25014 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.default-wasm (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8508 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12029 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10316 "Built successfully") | | |

### CODE REVIEW

**Comment by pxlcoder** [comment]
The style errors match existing code.

File: Source/WebCore/css/CSSValueKeywords.in [hunk 1]
```diff
@@ -1414,6 +1414,14 @@ spring
 steps
 apple-invert-lightness
 
+// Non-standard filters
+apple-blur enable-if=HAVE_CORE_MATERIAL
```
Comment (nt1m) [inline Source/WebCore/css/CSSValueKeywords.in]: Should be -apple- (apple-invert-lightness was a mistake)

Comment (pxlcoder) [reply Source/WebCore/css/CSSValueKeywords.in]: I originally had `-apple-blur`, but it wasn't working. Turns out that is due to `isAppleLegacyCSSValueKeyword`, which changes that into `-webkit-blur` during parsing.

I'll rename to `-apple-system-blur`, as the use of `system` will cause `isAppleLegacyCSSValueKeyword` to return false, and the keyword won't be unexpectedly transformed.

**Comment by mattwoodrow** [comment]
Having `-apple-system-blur` be a filter type implies (to me, at least) that there's ordering between that and any other filters provided. I.e `backdrop-filter: -apple-system-blur(-apple-system-material-thick) invert(1)` is different from `backdrop-filter: -apple-system-blur(-apple-system-material-thick) invert(1)`.

Based on the impl, this seems more like a non-ordered property of the backdrop layer. Do you think it'd be worth having a separate property for it instead?

**Comment by pxlcoder** [comment]
> Having `-apple-system-blur` be a filter type implies (to me, at least) that there's ordering between that and any other filters provided. I.e `backdrop-filter: -apple-system-blur(-apple-system-material-thick) invert(1)` is different from `backdrop-filter: invert(1) -apple-system-blur(-apple-system-material-thick)`.
> 
> Based on the impl, this seems more like a non-ordered property of the backdrop layer. Do you think it'd be worth having a separate property for it instead?

Makes sense. Thinking of adding `-apple-material` as a new property.

**Comment by pxlcoder** [comment]
Note that I ended up going with `-apple-visual-effect`, since this will be used for blur and vibrancy effects. And matches the platform `{UI/NS}VisualEffectView`.

**Review by aprotyas — APPROVED** [review]

**Review by whsieh — APPROVED** [review]

File: Source/WebCore/css/CSSValueKeywords.in [hunk 2]
```diff
@@ -1414,6 +1414,21 @@ spring
...
+-apple-system-blur-material-thick enable-if=HAVE_CORE_MATERIAL
+-apple-system-blur-material-chrome enable-if=HAVE_CORE_MATERIAL
+-apple-system-vibrancy-label enable-if=HAVE_CORE_MATERIAL
+-apple-system-vibrancy-secondary-label enable-if=HAVE_CORE_MATERIAL
+-apple-system-vibrancy-tertiary-label enable-if=HAVE_CORE_MATERIAL
+-apple-system-vibrancy-quaternary-label enable-if=HAVE_CORE_MATERIAL
+-apple-system-vibrancy-fill enable-if=HAVE_CORE_MATERIAL
+-apple-system-vibrancy-secondary-fill enable-if=HAVE_CORE_MATERIAL
+-apple-system-vibrancy-tertiary-fill enable-if=HAVE_CORE_MATERIAL
+-apple-system-vibrancy-separator enable-if=HAVE_CORE_MATERIAL
```
Comment (nt1m) [inline Source/WebCore/css/CSSValueKeywords.in:1430]: Not sure there's much point prefixing with `-system` (all of them are system values) or even `-apple` given the property name. I don't care too much though.

**Review by nt1m — APPROVED** [review]

File: Source/WebCore/css/CSSValueKeywords.in (hunk 2 above)
Comment (pxlcoder) [reply Source/WebCore/css/CSSValueKeywords.in:1430]: Going to keep this as-is for now, but we can re-consider naming as internal clients can't access this yet.

**Pull request merged**